### DISCUSSION TITLE
Integration of cloud clock with local FREMO RUT clocks

### DISCUSSION THREAD

**Discussion by fjallemark** [tellurianinteractive/Tellurian.Trains.ModuleMeetingApp#44] [Ideas] [answered]
This is a discussion of how to use the clock-server as master for FREMO RUT clocks.

### Background
RUT, short for Network-Clock-Telephone, used for Model railway purpose. Clocks in FREMO are similar to industrial Clock-Networks.
Several Slave-Clocks are driven by one Master-Clock, which in this case would be the *cloud server clock*.

### Rationale
The main purpose for this integration is that some of us like the real analogue clocks. If such clocks could have the clock-server as master, it will be possible to combine usage of analogue clocks and the clock web client app at the same meeting. For instance, the clock administrator will be able to control all clocks from a mobile phone. 

The use of analogue clocks should be combined with at least one big screen monitor, to be able to display information not possible to show on an analogue clock, for instance *weekday of operation*, *reason for stopping clock*, *who stopped the clock* and *free text messages*. 

### Controlling analogue clocks 
Master should delivers alternating positive and negative pulses for each second. Positive pulses makes the clock jump to even minute and negative pulses to odd minutes. Because this is a fast clock, the pulses will also be delivered in a faster pace that normal clocks. 
It is also necessary to *fast forward* clocks to a specified start time.

This is controlled by some hardware generating pulses, and the hardware should be controlled by some software. 

### Hardware requirements
The hardware must meet these requirements:
- Deliver pulses in correct voltage and with enough current to provide the analogue clocks with pulses.
- The pulses polarity and duration must be controlled from low current I/O signals from a device running the software.
- A question is if also zero pulse voltage must be supported, for instance when the clocks are stopped.
- The owners of each analogue clock must have a procedure to initially/manually synch the analogue clocks to a common start time.

### Software requirements
The software could be divided in two parts: the business logic and the I/O logic.

Requirements for the business logic are:
- Repeatedly polling the cloud clock for current game time and status.  
- If cloud clock is running real time, game clocks should be halted.
- Each time the minute changes, an event *minute changed* should be raise with data about the hour, minute.
- If time jumps more than a minute, for example when resetting clock to restart game, 
the software should generate faster pulses to advance the analogue clock the minutes required to be in synch with the new time.
It should be configurable the maximum speed clocks can be fast-forwarded.
- A separate logic determines how changes in minutes and translates to positive/negative pulses or zero.

The requirement for the I/O logic are:
- The API calls for current time are directed to the endpoint where the clock-server is running. This requires http/https over wired or wireless network. If calling the cloud-server, Internet access is also required.
- The I/O-logic uses data about positive/negative pulses or zero to activate I/O pins accordingly, which then controls the hardware.

### Handling actual I/O
There are several ways that the I/O logical signals created by the software are mapped to physical outputs.
Here some examples:
- RTS/CTS signals on a RS-232 interfaces.
- GPIO-pins on a Raspberry PI

**Comment by dbornawidacom** [comment]
### Handling via WLAN and esp32, based on wiFRED communication 

![image](https://user-images.githubusercontent.com/53570185/151845551-a2c76056-1fa3-4550-9db4-aa2fb4a3da75.png)

**Reply by fjallemark** [reply]
That looks good if there already exist a hardware with a WEB API for http-requests.  I have a number of questions:
1. Can you provide any documentation of the **esp32** WEB API?
2. What is the fastest speed (pulse frequency) that analogue clocks can handle?
3. When a http-request to set a positive voltage, does the **esp32** hold that voltage until another http-request is sent?
4. How does the **esp32** handles repeated http-requests for *positive* voltage (no *negative* in between)?

**Reply by dbornawidacom** [reply]
Hi,
this is just an idea of mine; there is no ready-made solution yet. My current project is to build null serie (200 pieces) of wiFreds from Heiko Rosemann. 
In this project i am developing a configurator in delphi (in future i would like to switch to visual studio :-)) to simplify managing the wiFreds. For this I extended the firmware and learned something about the programming of the esp32.
The wiFred looks
![image](https://user-images.githubusercontent.com/53570185/151850295-19198c83-9d48-4ce2-855d-319ebbf6ff8b.png)
My idea is to take the board of wiFred and replace 2 LED or switches with 2 relais to genereate the pulses on rut. Than i could test new rut interface. Later we can develop and produce its own board. 
I try to answer your questions:
1. You can see to the code of heiko: https://newheiko.github.io/wiFred/documentation/docu_en.html
2. Very slow. It may be depends on count of clocks in the line. Calc with 1 puls per second. In slave-clock-controler. this shoud be configured. If this is to short, some minutes will be lost.
3. I would use 2 pins in esp to connect to 2 relais (one +, one - ). So the rut is totaly separated. I think of a request like: http:\blabla\DoPuls?count=60. Esp will give 60 puls = minutes to rut-line 
4. esp32 is like a arduino with integrated WLAN

**Reply by newHeiko** [reply]
Interestingly I've created a somewhat similar project for WiFi clocks: https://github.com/newHeiko/WiFi-JMRI-Clock

The documentation is way behind, so let me give a quick introduction and share some pictures:

- Works with 2xAA batteries and an ESP12-module
- Designed to drive (two) cheap wall clocks as described here: http://www.jagsttalbahn-modelle.de/betrieb/modellzeit/modellzeit.html (not my page) - although I never managed to remove the second hand and get the clock work reliably on a minute tick, so my clocks have a second hand. The firmware does support other options including a sweeping hand.
- Gets the time from a (JMRI) wiThrottle server (could probably easily be adopted to any other non-encrypted TCP based protocol)
- Keeps "local" and "network" time so the analog clock can catch up with maximum speed when the network time jumps.

I still have a couple of blank PCBs, in case anyone wants to try some SMD (0805 and larger) soldering, and I'd even throw in some of those SMD pushbutton switches.

For RUT, it would obviously need a "booster". And if there's 12V/24V available anyway, maybe it would make more sense to create something based on a NodeMCU or similar development boards.

And here's the images:
![image](https://user-images.githubusercontent.com/35038740/157531618-2d6e4881-4f75-42ce-bf6c-43223b76f511.png)
![image](https://user-images.githubusercontent.com/35038740/157531651-5c417915-7722-4dc6-a8bc-f092f6dabdf3.png)
![image](https://user-images.githubusercontent.com/35038740/157530947-d1e978ef-73cc-4191-9e7e-91090f042f00.png)

We could also get together on Jitsi or Zoom if you want to know more.

Heiko

**Reply by newHeiko** [reply]
P.S: The wiFred code will (basically) also run on any other ESP32-S2 demo board, for example these: https://www.reichelt.de/entwicklungsboard-esp32-s2-wroom-esp32s2saola1m-p311736.html?&nbc=1 - once again it might make more sense to develop based on that rather than a full wiFred.

**Comment by freddepettersson** [comment]
Hi
I was involved developing a connection between the FastClock and RUT in 2013 Mr Daniel Bergkvist made a Jave application to run on a Vellermann I/O card. It worked nice ( we hade some issues with disturbance on the signals due to the load of secondary puls clock using a rather high Amp). He also made it work using a USB-Serial port adapter, he used RX / TX signal. I know he made a simple hardware that was tested ok during a meeting.

The basic setting was:
- The client uses the master signal from a digital FastClock server
- one of the parameters you needed to typed in was the analog clock start
position. The client found the signal from the digital clock and its
starting time. The client compared the set positions.
- then it was just to press "start clock". The client compares the "play
time" and then its start to catch up.

We found out that the signal 12/24 VDC to the analog clocks needed a pulse time and a pause time (delay).
We found out and tested with 6 analog clocks that a pulse setting around 1,5 - 2.0 seconds and a short delay of 0,5 sec was a nice setup to get all analog clocks to respond when speeding to a set time .
The minute pulse changed polarity of course and next minute's step was
triggered with a new puls of 2.0 seconds , then pause etc.....

The same timers was used when the playtime runned , but of course delay was longer due to the requested speed selected in the FastClock server

[image: image.png]

Regards
Fredrik

**Reply by freddepettersson** [reply]
![image](https://user-images.githubusercontent.com/80942768/152023449-3927c42b-01e6-48b4-bfba-f8340cc615bc.png)

**Comment by dbornawidacom** [comment]
**What about Second hand without time reduction on browser clock**

So that the second hand doesn't race like crazy, we could proceed as follows:
With a time reduction of 1:6, a model minute lasts 10 seconds. At the 59th second the hand should wait a second.
We can do as follows
Start at 0 second
![image](https://user-images.githubusercontent.com/53570185/152641887-7c1ab25e-fc69-4952-a3f6-bade40570c62.png)
 
1. Second go to 1
![image](https://user-images.githubusercontent.com/53570185/152641907-601b7849-1a8d-4145-a1d5-aeaeb6591e85.png)

2. Second go to 2
![image](https://user-images.githubusercontent.com/53570185/152641914-f047d153-111a-4720-b6f1-0468122b99b0.png)

3. Second go to 54
![image](https://user-images.githubusercontent.com/53570185/152641922-55ef7219-9d4e-4e22-981c-de0586ca3bc4.png)

4. Second go to 55
5. Second go to 56
6. Second go to 57
7. Second go to 58
8. Second go to 59
 ![image](https://user-images.githubusercontent.com/53570185/152641935-561ca066-0fc4-4005-b52c-a55c1e3a0f79.png)

9. Second do nothing
10. Second go to 0 next minute
![image](https://user-images.githubusercontent.com/53570185/152641946-4728547f-d4a4-49ca-b6a1-bfe21c6de6c8.png)

With a reduction of 1:3, a model minute  has 20 seconds. So we have to jump to 44. 
So the second hand will move with real speed.

**Reply by freddepettersson** [reply]
Hi Stefan
I think the second hand will be more confusing than helpull in this case. 
I understand that it would look nice , but to have a secondhand jumping on the clock will make a distraction.

/Fredrik

**Comment by fjallemark** [comment]
Today, I have released the souce-code for translating API-calls to the cloud clock to logical pulses for driving an analogue clock. This means that anyone that wants to create a hardware, now can get the data when to create a positive or negative pulse. The application also automatically fast forward the pulsing to the current time if they get out of sync. Take a look at the [source code](https://github.com/fjallemark/ClockPulseService). One next step is to implement a *pulse sink* that broadcast the pulses on a local network, where hardware drivers can listen and create the pulses needed for moving the clock.

**Comment by fjallemark** [comment, marked as answer]
Today, I have got a **Raspberry Pi** with a [**Rpi Relay Board**](https://www.waveshare.com/wiki/RPi_Relay_Board) working to generate electric impulses to control an analogue clock. It is of course possible to use the I/O pins on the Raspberry Pi directly without the relay board.

**Comment by fjallemark** [comment]
There is now a working and tested solution for integrating analogue clocks with the fast clock app.
